FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,49 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let managedCache = try ManagedCache.newInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.reset()
```

Comment:
`context.reset` invalidates all the managed objects of this context which in turn can be an issue if some of these objects are in use. To only reset the changes `context.rollback` needs to be used. 

Please replace `reset` call with `rollback`

```suggestion
				context.rollback()
```

Reply:
done

Reply:
Please make sure this change is pushed, I am still seeing `context.reset`